### PR TITLE
feat(workspaces): handle collection sub tab from the link COMPASS-7731

### PR DESCRIPTION
**Pull request by mabaasit**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-collection/src/components/collection-tab.tsx [hunk 1]
```diff
@@ -66,6 +66,22 @@ const TabTitleWithStats = ({
...
+   */
+  initialAggregation?: unknown;
+  /**
+   * Initial aggregation pipeline to set in the agg builder
+   */
+  initialPipeline?: unknown[];
+  /**
+   * Initial stringified aggregation pipeline to set in the agg builder
+   */
+  initialPipelineText?: string;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-tab.tsx:84]: This is already part of `CollectionTabOptions`

File: packages/compass-collection/src/stores/collection-tab.ts [hunk 2]
```diff
@@ -12,28 +12,17 @@ import type Collection from 'mongodb-collection-model';
...
-   */
-  initialAggregation?: unknown;
-  /**
-   * Initial aggregation pipeline to set in the agg builder
-   */
-  initialPipeline?: unknown[];
-  /**
-   * Initial stringified aggregation pipeline to set in the agg builder
-   */
-  initialPipelineText?: string;
```
Comment (gribnoysup) [inline packages/compass-collection/src/stores/collection-tab.ts:36]: We should't remove those from tab options, these are different use-cases to a subtab

File: packages/compass-collection/src/modules/collection-tab.ts [hunk 3]
```diff
@@ -32,12 +33,7 @@ export type CollectionState = {
     | 'free_storage_size'
   > | null;
   metadata: CollectionMetadata | null;
-  currentTab:
-    | 'Documents'
-    | 'Aggregations'
-    | 'Schema'
-    | 'Indexes'
-    | 'Validation';
+  currentTab: CollectionSubtab;
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/collection-tab.ts:36]: The source of truth for this is now workspaces plugin, so we should remove this from the store in the collection tab otherwise we are regressing back to duplicating workspaces state in the application

File: packages/compass-collection/src/stores/collection-tab.spec.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ const defaultMetadata = {
 
 const defaultTabOptions = {
   namespace: defaultMetadata.namespace,
+  subTab: 'Documents',
```
Comment (gribnoysup) [inline packages/compass-collection/src/stores/collection-tab.spec.ts:21]: This should be `as const` to avoid doing unneccessary as any cast below (it's always a good practice to limit those to a minimum and actually resolve TS errors if possible)

```suggestion
  subTab: 'Documents' as const,
```

File: packages/compass-collection/src/stores/collection-tab.ts [hunk 5]
```diff
@@ -100,15 +76,15 @@ export function activatePlugin(
...
 
   on(localAppRegistry, 'open-create-search-index-modal', () => {
-    store.dispatch(selectTab('Indexes'));
+    store.dispatch(selectTab(CollectionSubtabs.Indexes));
   });
 
   on(localAppRegistry, 'generate-aggregation-from-query', () => {
-    store.dispatch(selectTab('Aggregations'));
+    store.dispatch(selectTab(CollectionSubtabs.Aggregations));
   });
```
Comment (gribnoysup) [inline packages/compass-collection/src/stores/collection-tab.ts:88]: Now that this state lives in workspaces plugin, we need to remove this logic from this store

File: packages/compass-workspaces/src/types.ts [hunk 6]
```diff
@@ -18,6 +20,7 @@ export type CollectionsWorkspace = {
 export type CollectionWorkspace = {
   type: 'Collection';
   namespace: string;
+  subTab: CollectionSubtab;
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/types.ts:23]: Can we type this as string union instead of a enum? Makes a bit easier to integrate and means we don't need to re-export this type from compass-web so that mms can use it (and easier for us to use cross plugins)

File: packages/compass-workspaces/src/stores/workspaces.spec.ts [hunk 7]
```diff
@@ -73,7 +73,11 @@ describe('tabs behavior', function () {
       const store = configureStore();
       openTabs(store);
       store.dispatch(
-        openWorkspace({ type: 'Collection', namespace: 'test.bar' })
+        openWorkspace({
+          type: 'Collection',
+          namespace: 'test.bar',
+          subTab: 'Indexes',
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/stores/workspaces.spec.ts:79]: This doesn't look right, this should be optional when opening a workspace, you can't provide a required subTab and pass an initial query / agg at the same time

File: packages/compass-collection/src/modules/collection-tab.ts (hunk 3 above)
Comment (mabaasit) [reply packages/compass-collection/src/modules/collection-tab.ts:36]: Yes, I was removing this as part of communicating the (tab) change back to the workspaces.

File: packages/compass-collection/src/stores/collection-tab.ts (hunk 5 above)
Comment (mabaasit) [reply packages/compass-collection/src/stores/collection-tab.ts:88]: In this case, these events will only be applicable to the active namespace, right?

File: packages/compass-workspaces/src/types.ts (hunk 6 above)
Comment (mabaasit) [reply packages/compass-workspaces/src/types.ts:23]: yeah no strong preference here. will make the change

File: packages/compass-collection/src/stores/collection-tab.ts (hunk 5 above)
Comment (gribnoysup) [reply packages/compass-collection/src/stores/collection-tab.ts:88]: They should stay attached to the tab local app registry scope. If we're moving the state up, this generally means that we should just set up the subscription when the tab is first created

File: packages/compass-workspaces/src/stores/workspaces.spec.ts (hunk 7 above)
Comment (mabaasit) [reply packages/compass-workspaces/src/stores/workspaces.spec.ts:79]: We use `useWorkspace` hook to expose `openCollectionWorkspace` and `openEditViewWorkspace`. These methods now contain the logic to decide which subtab to open, which was previously in `collection-tab.ts`. With that, I think it should be required for this method and optional for `openCollectionWorkspace`.

File: packages/compass-collection/src/stores/collection-tab.ts (hunk 5 above)
Comment (mabaasit) [reply packages/compass-collection/src/stores/collection-tab.ts]: Added in [cec4882](https://github.com/mongodb-js/compass/pull/5568/commits/cec4882641fbe89770c4e5565f41b7d05a729022)

File: packages/compass-workspaces/src/provider.tsx [hunk 8]
```diff
@@ -65,7 +65,7 @@ export type WorkspacesService = {
     options?: TabOptions &
       Omit<
         Extract<OpenWorkspaceOptions, { type: 'Collection' }>,
-        'type' | 'namespace' | 'editViewName'
+        'type' | 'namespace' | 'editViewName' | 'subTab'
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/provider.tsx:68]: For the openWorkspaceOptions this should probably be initialSubTab and should be optionally allowed

File: packages/compass-workspaces/src/provider.tsx [hunk 9]
```diff
@@ -150,10 +150,20 @@ export const WorkspacesServiceProvider: React.FunctionComponent<{
         );
       },
       openCollectionWorkspace: (namespace, options) => {
+        const isAggregationsSubtab = Boolean(
+          options?.initialAggregation ||
+            options?.initialPipeline ||
+            options?.initialPipelineText
+        );
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/provider.tsx:157]: I don't think it's great that we're spreading the logic of opening a tab in multiple palces, I'd have a preference for these re-exported methods to be very thin wrappers over the openWorkspace action. We should have a single place that we can trust to always open the tab consistently across the application and inside this plugin

File: packages/compass-workspaces/src/stores/workspaces.ts [hunk 10]
```diff
@@ -361,30 +366,10 @@ const reducer: Reducer<WorkspacesState> = (
...
-        if (id === state.activeTabId) {
-          newActiveTabId = newTab.id;
-        }
-        return newTab;
-      }
-      return tab;
-    });
-    if (tabsRenamed === 0) {
-      return state;
-    }
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/stores/workspaces.ts:383]: Why is this removed from there? Reducer shoud be deriving state from previous state and actions as per [redux guidelines](https://redux.js.org/style-guide#put-as-much-logic-as-possible-in-reducers)

File: packages/compass-collection/src/modules/collection-tab.ts [hunk 11]
```diff
@@ -4,22 +4,22 @@ import type Collection from 'mongodb-collection-model';
...
     localAppRegistry: AppRegistry;
     dataService: DataService;
+    workspaces: ReturnType<typeof workspacesServiceLocator>;
   },
-  Action
+  A
 >;
 
 export type CollectionState = {
+  workspaceTabId: string;
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/collection-tab.ts:22]: Do we have a ticket for changing this to use context value? Would be good to have it here in a TODO

**Review by gribnoysup — APPROVED** [review]
Awesome work!

File: packages/compass-collection/src/modules/collection-tab.ts (hunk 11 above)
Comment (mabaasit) [reply packages/compass-collection/src/modules/collection-tab.ts:22]: will link it! thanks!

File: packages/compass-workspaces/src/stores/workspaces.ts [hunk 12]
```diff
@@ -403,6 +429,26 @@ const reducer: Reducer<WorkspacesState> = (
...
+      tabs: state.tabs.map((tab) => {
+        if (tab.id === action.tabId) {
+          return {
+            ...tab,
+            subTab: action.subTab,
+          };
+        }
+        return tab;
+      }),
+    };
```
Comment (gribnoysup) [inline packages/compass-workspaces/src/stores/workspaces.ts:449]: One small thing caught my eye here on another look. You probably want to short circuit if the subtab is already set to the new value, otherwise you are unnecessarily recreate both tab array and the tab state itself when nothing actually changed. Something like:

```
if (tabToChange.subTab === action.subTab) {
  return state;
}
// ... then your logic
```

Comment (mabaasit) [reply packages/compass-workspaces/src/stores/workspaces.ts:453]: Just added in [b653813](https://github.com/mongodb-js/compass/pull/5568/commits/b65381312154ef6138f69e08f7eae752e0be1c8c)

File: packages/compass-workspaces/src/types.ts [hunk 13]
```diff
@@ -16,8 +23,14 @@ export type CollectionsWorkspace = {
 };
 
 export type CollectionWorkspace = {
+  // TODO: COMPASS-7782 use hook to get the tab id within workspace.
```
Comment (addaleax) [inline packages/compass-workspaces/src/types.ts:26]:
```suggestion
  // TODO(COMPASS-7782): use hook to get the tab id within workspace.
```

Let's stick to the format we use everywhere else :)

File: packages/compass-schema/src/index.ts [hunk 14]
```diff
@@ -35,6 +35,6 @@ export const CompassSchemaHadronPlugin = registerHadronPlugin<
   }
 );
 export const CompassSchemaPlugin = {
-  name: 'Schema',
+  name: 'Schema' as const,
```
Comment (addaleax) [inline packages/compass-schema/src/index.ts:38]: 💙

File: packages/compass-workspaces/src/stores/workspaces.ts [hunk 15]
```diff
@@ -455,40 +505,48 @@ export const openWorkspace = (
...
-              type: WorkspacesActions.FetchCollectionTabInfo,
-              namespace: workspaceOptions.namespace,
-              info: {
-                isTimeSeries: coll.isTimeSeries,
-                isReadonly: coll.readonly ?? coll.isView,
-                sourceName: coll.sourceName,
-              },
+    if (workspaceOptions.type === 'Collection') {
+      if (!getState().collectionInfo[workspaceOptions.namespace]) {
+        // Fetching extra meta for collection should not block tab opening
```
Comment (addaleax) [inline packages/compass-workspaces/src/stores/workspaces.ts:510]:
```suggestion
        // Fetching extra metadata for collection should not block tab opening
```

File: packages/compass-workspaces/src/stores/workspaces.ts [hunk 16]
```diff
@@ -455,40 +505,48 @@ export const openWorkspace = (
...
+                },
+              });
+            }
+          } catch (err) {
+            // It's okay if we failed to fetch this optional metadata, this error
+            // can be ignored
+            if ((err as Error).name === 'MongoServerError') {
+              return;
+            }
+            throw err;
```
Comment (addaleax) [inline packages/compass-workspaces/src/stores/workspaces.ts:537]: Should this really end up being an unhandled rejection? Logging the error (maybe even in both the `MongoServerError` and non-`MongoServerError`) might be a better choice here

Comment (mabaasit) [reply packages/compass-workspaces/src/stores/workspaces.ts:537]: I am fine with adding that and removing this. Nonetheless, I'd let @gribnoysup comment on this as this was here before.

Comment (gribnoysup) [reply packages/compass-workspaces/src/stores/workspaces.ts:537]: I think we were talking at some point about not swallowing all the errors if we don't expect them, but can't say I remember arriving on a decision, so feel free to remove the throw. Adding logging make sense

**Review by addaleax — APPROVED** [review]
